### PR TITLE
Lab6

### LINKED ISSUE #566: Lab 6 - puppy storer

### PR DESCRIPTION
**Pull request by nicko-lee**
Fixes #566 

Review of colleague's PR #342 

Implement puppy storer interface with mapStore and syncStore and write tests.

### CODE REVIEW

**Review by juliaogris — CHANGES_REQUESTED** [review]
Please fix failing build before adding `Ready for review` label.
Thanks!

**Comment by kasunfdo** [comment]
@nicko-lee , add main.go. Without that, travis build fails.

File: 06_puppy/nickolee/mapStore.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package puppystorer
+
+import "fmt"
+
+// IncrementCounter increases the ID counter everytime a new Puppy is created to prevent overwrite issues
+// in DeletePuppy()
+func (ms *MapStore) IncrementCounter() {
```
Comment (juliaogris) [inline 06_puppy/nickolee/mapStore.go]: Make private or don't bother with it at all (it's a one liner).
Incrementing Counter is not something Store or MapStore should expose (it's more of an implementation detail).
In a sql for instance you'd use SERIAL or auto-increment and not have a code equivalent.

File: 06_puppy/nickolee/storerInterface_test.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
+package puppystorer
```
Comment (juliaogris) [inline 06_puppy/nickolee/storerInterface_test.go]: i think it's somewhat established to have go filenames all lowercase (same below).
storer_test.go should suffice - `er` usually indicates interface (e.g. Reader, Writer, Closer....)

File: 06_puppy/nickolee/types.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+type Storer interface {
+	CreatePuppy(puppy *Puppy) int // takes an address as an argument which makes sense since you are modifying that object
+	ReadPuppy(id int) (*Puppy, error)
+	UpdatePuppy(id int, puppy *Puppy) error
+	DeletePuppy(id int) (bool, error)
+}
+
+// type MapStore map[int]Puppy
+// This will serve as our in-memory DB
+type MapStore struct {
```
Comment (juliaogris) [inline 06_puppy/nickolee/types.go]: put with `mapstore.go`

File: 06_puppy/nickolee/types.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	store   map[int]Puppy
+	counter int
```
Comment (juliaogris) [inline 06_puppy/nickolee/types.go]: would call this nextID (that's how it's used) and increment it after asignement.

File: 06_puppy/nickolee/types.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+// SyncStore struct. To serve as alternative in-memory DB. It also implements Storer interface
+type SyncStore struct {
```
Comment (juliaogris) [inline 06_puppy/nickolee/types.go]: same as MapStore

File: 06_puppy/nickolee/mapStore.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	if puppy, ok := ms.store[id]; ok {
+		return &puppy, nil
+	}
+	// else return nil pointer to puppy and an error
+	return nil, fmt.Errorf("puppy with ID: %d does not exist", id)
+}
+
+// UpdatePuppy lets you update a "row" in MapStore
+func (ms *MapStore) UpdatePuppy(id int, puppy *Puppy) error {
+	if _, ok := ms.store[id]; !ok {
```
Comment (juliaogris) [inline 06_puppy/nickolee/mapStore.go]: what if puppy.ID and id don't match? you should add at least the error check for this, better just use Puppy. No need to pass in ID twice.

File: 06_puppy/nickolee/mapStore.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	return nil
+}
+
+// DeletePuppy lets you delete a specific puppy in MapStore
+func (ms *MapStore) DeletePuppy(id int) (bool, error) {
+	if _, ok := ms.store[id]; !ok {
+		return false, fmt.Errorf("puppy with ID: %d does not exist", id)
+	}
+	delete(ms.store, id)
+	return true, nil
```
Comment (juliaogris) [inline 06_puppy/nickolee/mapStore.go]: no need for bool _and_ error, just error should suffice.

File: 06_puppy/nickolee/storerInterface_test.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+)
+
+type StorerTest struct {
```
Comment (juliaogris) [inline 06_puppy/nickolee/storerInterface_test.go]: s/StorerTest/StorerSuite

File: 06_puppy/nickolee/storerInterface_test.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	suite.False(deleted)
+	suite.Error(err) // assert there should be an error cos puppy with id 7 doesn't exist
+}
+
+// Run tests for Storer using MapStore implementation of Storer interface
+func TestMapStore(t *testing.T) {
+	s := StorerTest{
+		store: &MapStore{store: make(map[int]Puppy), counter: 0},
+	}
+	suite.Run(t, &s)
```
Comment (juliaogris) [inline 06_puppy/nickolee/storerInterface_test.go]: better IMO (same below)
```go
suite.Run(t, &StorerSuite{NewMapStore()})
```

File: 06_puppy/nickolee/syncStore.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
+package puppystorer
+
+import "fmt"
+
+// IncrementCounter increases the ID counter everytime a new Puppy is created to prevent overwrite issues
+// in DeletePuppy()
+func (m *SyncStore) IncrementCounter() {
```
Comment (juliaogris) [inline 06_puppy/nickolee/syncStore.go]: same as mapstore.

File: 06_puppy/nickolee/syncStore_test.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+		{Breed: "Simba", Colour: "Pumba", Value: 92300.90},
+		{Breed: "Direwolf", Colour: "Isn't technically a wolf", Value: 22300.90},
+	}
+
+	for _, testPuppy := range tests {
+		ss.CreatePuppy(testPuppy)
+		retrievedPuppy, _ := ss.ReadPuppy(testPuppy.ID)
+		assert.Equalf(testPuppy, retrievedPuppy, "Testcase TestCreateAndRetrievePuppySyncStore failed")
+	}
+
```
Comment (juliaogris) [inline 06_puppy/nickolee/syncStore_test.go]: unnecessary blank line.

File: 06_puppy/nickolee/syncStore_test.go [hunk 12]
```diff
@@ -0,0 +1,29 @@
+package puppystorer
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+/// Writing one test to give confidence that mapstore is retrieving puppies correctly
+func TestCreateAndRetrievePuppySyncStore(t *testing.T) {
```
Comment (juliaogris) [inline 06_puppy/nickolee/syncStore_test.go]: why not put this with the Suite?

**Review by juliaogris — DISMISSED** [review]
a few change requests inline.
sorry for the lack of politeness around it all - just too many reviews to get through.
my utmost respect for still sticking with it - many comments are personal opinion. fee free to give push back.

File: 06_puppy/nickolee/syncStore_test.go (hunk 12 above)
Comment (anzboi) [reply 06_puppy/nickolee/syncStore_test.go]: I had a look at this with @nicko-lee. He didn't have these tests to begin with. All tests were passing and had full coverage.
Unfortunately there was a bug in the way he was calculating the next ID (using the length of the current internal map). This cause the mapstore to overwrite old entries when you did `Create`, `Delete`, `Create`. I encouraged him to add this test and work on fixing the ID generation, and make sure it stays fixed with this test.

If you think this is overkill, I wouldn't bother with these tests. It just came out of the discussion that just using an interface test can result in 100% coverage, all tests pass, but still have fundamentally broken code.

File: 06_puppy/nickolee/mapStore.go (hunk 1 above)
Comment (nicko-lee) [reply 06_puppy/nickolee/mapStore.go]: Thanks for the tip. Done

File: 06_puppy/nickolee/types.go (hunk 3 above)
Comment (nicko-lee) [reply 06_puppy/nickolee/types.go]: Done

File: 06_puppy/nickolee/types.go (hunk 4 above)
Comment (nicko-lee) [reply 06_puppy/nickolee/types.go]: Thanks for the tip. Have renamed.

File: 06_puppy/nickolee/storerInterface_test.go (hunk 8 above)
Comment (nicko-lee) [reply 06_puppy/nickolee/storerInterface_test.go]: Done

File: 06_puppy/nickolee/storerInterface_test.go (hunk 9 above)
Comment (nicko-lee) [reply 06_puppy/nickolee/storerInterface_test.go]: Done and created NewMapStore() and NewSyncStore() functions

File: 06_puppy/nickolee/syncStore_test.go (hunk 11 above)
Comment (nicko-lee) [reply 06_puppy/nickolee/syncStore_test.go]: Easily fixed :)

File: 06_puppy/nickolee/mapStore.go (hunk 6 above)
Comment (anzboi) [reply 06_puppy/nickolee/mapStore.go]: I talked with @nicko-lee about this. We decided to ignore the id in the puppy. This allows code like
```
s.UpdatePuppy(1000, &Puppy{Name: "Fido", Breed: "Poodle", Value: 1234})
```

This feels like it should work, whereas erroring out on id clashes will cause this to throw an error.

**Review by anzboi — APPROVED** [review]
LGTM

**Pull request merged**